lib/mongo/server/connection.rb

```diff
@@ -155,7 +155,15 @@ def ping
 
       def deliver(messages)
         write(messages)
-        messages.last.replyable? ? read : nil
+
+        if messages.last.replyable?
+          # Protection against returning the response to a previous request. See
+          # RUBY-1117
+          reply = read until reply && reply.response_to == messages.last.request_id
```

Comment:
Ok I actually found a not sexy but very simple and easy fix, as thread interruption calls `ensure` block but not `rescue` (that's why the socket is checked but not closed), we simply need to use and ensure block to close the socket. That would change:
```ruby
def ensure_connected
  ensure_same_process!
  connect!
  begin
    yield socket
  rescue Exception => e
    disconnect!
    raise e
  end
end
```
To:
```ruby
def ensure_connected
  ensure_same_process!
  connect!
  begin
    res = yield socket
    success = true
    res
  ensure
    success or disconnect!
  end
end
```
We confirmed this with our example and our spec suite, it properly closes the socket in case of both `Exception` AND `Thread.kill`. This way the dirty socket is never reused. We should of course keep the UnexpectedResponse exception because it protects against regressions or other bugs, but this covers the problem better than simply raise randomly or forcing you to change how you write ruby code. WDYT?